chain/store/stateGenerator.go

```diff
@@ -7,239 +7,187 @@ import (
 
 	"github.com/nknorg/nkn/block"
 	"github.com/nknorg/nkn/chain"
-	. "github.com/nknorg/nkn/common"
+	"github.com/nknorg/nkn/common"
 	"github.com/nknorg/nkn/crypto"
 	"github.com/nknorg/nkn/pb"
 	"github.com/nknorg/nkn/program"
 	"github.com/nknorg/nkn/transaction"
 	"github.com/nknorg/nkn/util/config"
 )
 
-func (cs *ChainStore) spendTransaction(states *StateDB, txn *transaction.Transaction, totalFee Fixed64, genesis bool, height uint32) error {
+func (cs *ChainStore) spendTransaction(states *StateDB, txn *transaction.Transaction, totalFee common.Fixed64, genesis bool, height uint32) error {
 	pl, err := transaction.Unpack(txn.UnsignedTx.Payload)
 	if err != nil {
 		return err
 	}
 
+	pg, err := txn.GetProgramHashes()
+	if err != nil {
+		return err
+	}
+
+	if txn.UnsignedTx.Fee > 0 {
+		if err = states.UpdateBalance(pg[0], config.NKNAssetID, common.Fixed64(txn.UnsignedTx.Fee), Subtraction); err != nil {
```

Comment:
Here reimplement to balance subtract `Tx.Fee` & `payload.Amount` individual.
Could statesDB revert to previous state if sub `payload.Amount` met err later?

Reply:
It won't. DB is changed in batch. Before .Commit() is called, no changes will be persisted

Reply:
That is OK.